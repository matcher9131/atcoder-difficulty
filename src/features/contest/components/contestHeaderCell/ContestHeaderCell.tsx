import clsx from "clsx";
import type { ReactNode } from "react";
import { cellClassNames } from "../../../../common/cellClassNames";

export type ContestHeaderCellProps = {
    readonly textColor: string;
    readonly displayName: string;
    readonly linkHref: string;
    readonly onStatsButtonClick: () => void;
    readonly statsIconHref: string;
};

export const ContestHeaderCell = ({
    textColor,
    displayName,
    linkHref,
    onStatsButtonClick,
    statsIconHref,
}: ContestHeaderCellProps): ReactNode => {
    return (
        <td className={clsx(cellClassNames, "gap-x-1")}>
            <span className={textColor}>â—‰</span>
            <a href={linkHref} target="_blank" rel="noreferrer" className="link link-hover link-primary truncate">
                {displayName}
            </a>
            <button onClick={onStatsButtonClick} className="hidden group-hover:block group-focus-within:block">
                <svg
                    role="img"
                    aria-label="Show graph"
                    className="h-6 w-6 fill-current cursor-pointer hover:invert-30 transition-all"
                >
                    <title>Show stats</title>
                    <use href={statsIconHref} />
                </svg>
            </button>
        </td>
    );
};
