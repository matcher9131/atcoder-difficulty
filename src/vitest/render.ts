import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
