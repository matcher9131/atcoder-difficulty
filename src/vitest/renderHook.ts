import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
