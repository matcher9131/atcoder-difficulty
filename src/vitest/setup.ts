import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
