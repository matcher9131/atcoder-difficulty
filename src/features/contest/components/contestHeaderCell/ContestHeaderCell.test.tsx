import { screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import { ContestHeaderCell } from "./ContestHeaderCell";
import { render } from "../../../../vitest/render";

// ContestHeaderCell.test.tsx

// Mock react-i18next
vi.mock("react-i18next", () => ({
    useTranslation: () => ({
        t: (key: string) => key,
    }),
}));

describe("ContestHeaderCell", () => {
    const defaultProps = {
        textColor: "text-red-500",
        displayName: "Sample Contest",
        linkHref: "https://example.com",
        onStatsButtonClick: vi.fn(),
        statsIconHref: "#icon-graph",
    };

    it("renders colored circle with correct class", () => {
        render(<ContestHeaderCell {...defaultProps} />);
        const circle = screen.getByText("â—‰");
        expect(circle).toHaveClass(defaultProps.textColor);
    });

    it("renders displayName as a link with correct href and attributes", () => {
        render(<ContestHeaderCell {...defaultProps} />);
        const link = screen.getByRole("link", { name: defaultProps.displayName });
        expect(link).toHaveAttribute("href", defaultProps.linkHref);
        expect(link).toHaveAttribute("target", "_blank");
        expect(link).toHaveAttribute("rel", "noreferrer");
        expect(link).toHaveClass("link-primary");
    });

    it("renders stats button and calls handler on click", () => {
        render(<ContestHeaderCell {...defaultProps} />);
        const button = screen.getByRole("button");
        fireEvent.click(button);
        expect(defaultProps.onStatsButtonClick).toHaveBeenCalled();
    });

    it("renders SVG with correct icon href and aria-label", () => {
        render(<ContestHeaderCell {...defaultProps} />);
        const svg = screen.getByRole("img", { name: "contestHeaderCell.showStatsButtonLabel" });
        expect(svg).toBeInTheDocument();
        const useElem = svg.querySelector("use");
        expect(useElem).toHaveAttribute("href", defaultProps.statsIconHref);
        expect(svg.querySelector("title")?.textContent).toBe("contestHeaderCell.showStatsButtonLabel");
    });

    it("matches snapshot", () => {
        const { container } = render(<ContestHeaderCell {...defaultProps} />);
        expect(container).toMatchSnapshot();
    });
});
