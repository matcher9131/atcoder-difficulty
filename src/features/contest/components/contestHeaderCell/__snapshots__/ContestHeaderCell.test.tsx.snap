// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContestHeaderCell > matches snapshot 1`] = `
<div>
  <td
    class="flex px-3 py-auto h-18 items-center border-t-1 border-base-content/75 group gap-x-1"
  >
    <span
      class="text-red-500"
    >
      â—‰
    </span>
    <a
      class="link link-hover link-primary truncate"
      href="https://example.com"
      rel="noreferrer"
      target="_blank"
    >
      Sample Contest
    </a>
    <button
      class="hidden group-hover:block group-focus-within:block"
    >
      <svg
        aria-label="contestHeaderCell.showStatsButtonLabel"
        class="h-6 w-6 fill-current cursor-pointer hover:invert-30 transition-all"
        role="img"
      >
        <title>
          contestHeaderCell.showStatsButtonLabel
        </title>
        <use
          href="#icon-graph"
        />
      </svg>
    </button>
  </td>
</div>
`;
